src/query_plan/fetch_dependency_graph.rs

```diff
@@ -292,18 +634,568 @@ impl DeferTracking {
+    // See the comment on the `optimize_sibling_typenames()` method to see why this exists.
+    if let Some(name) = operation.sibling_typename() {
+        // We need to add the query __typename for the current type in the current node.
+        // Note that `name` is the alias or '' if there is no alias
+        let alias = if name.is_empty() {
+            None
+        } else {
+            Some(name.clone())
+        };
+        let typename_field = Arc::new(OpPathElement::Field(NormalizedField::new(
+            NormalizedFieldData {
+                // TODO: should this be a subgraph schema?
+                schema: dependency_graph.supergraph_schema.clone(),
+                field_position: operation
+                    .parent_type_position()
+                    .introspection_typename_field(),
+                alias,
+                arguments: Default::default(),
+                directives: Default::default(),
+                sibling_typename: None,
+            },
+        )));
+        let typename_path = stack_item
+            .node_path
+            .path_in_node
+            .with_pushed(typename_field.clone());
+        let node =
+            FetchDependencyGraph::node_weight_mut(&mut dependency_graph.graph, stack_item.node_id)?;
+        node.selection_set_mut().add_at_path(&typename_path, None)?;
+        dependency_graph.defer_tracking.update_subselection(
+            &DeferContext {
+                path_to_defer_parent: Arc::new(
+                    stack_item
+                        .defer_context
+                        .path_to_defer_parent
+                        .with_pushed(typename_field),
+                ),
+                ..stack_item.defer_context.clone()
+            },
+            None,
+        )
+    }
+    let (Some(updated_operation), updated_defer_context) = extract_defer_from_operation(
+        dependency_graph,
+        operation,
+        &stack_item.defer_context,
+        &stack_item.node_path,
+    ) else {
+        return Err(FederationError::internal(format!(
+            "Extracting @defer from {operation:?} should not have resulted in no operation"
+        )));
+    };
+    let mut updated = ComputeNodesStackItem {
+        tree: &child.tree,
+        node_id: stack_item.node_id,
+        node_path: stack_item.node_path.clone(),
+        context: stack_item.context,
+        defer_context: updated_defer_context,
+    };
+    if let Some(conditions) = &child.conditions {
+        // We have @requires or some other dependency to create nodes for.
+        let (required_node_id, require_path) = handle_requires(
+            dependency_graph,
+            edge_id,
+            conditions,
+            (stack_item.node_id, &stack_item.node_path),
+            stack_item.context,
+            &updated.defer_context,
+            created_nodes,
+        )?;
+        updated.node_id = required_node_id;
+        updated.node_path = require_path;
```

Comment:
do we need to also extend `created_nodes` with `requires` nodes? https://github.com/apollographql/federation/blob/main/query-planner-js/src/buildPlan.ts#L4109

Reply:
In JS `createdGroups: FetchGroup[]` is part of the return value of `handleRequires()`. In Rust instead, `handle_requires` is still a TODO in this PR but it takes a `&mut IndexSet<NodeIndex>` parameter so it can insert directly instead of returning and intermediate set or vec.

Reply:
since it could be easily missed in subsequent PR -> can you add some comment/todo so we don't forget?

Reply:
I’ve added a `PORT_NOTE` comment